### PR TITLE
Cancel release on validation and manifest error

### PR DESCRIPTION
**Pull request by tomahawk28**
Toward https://github.com/giantswarm/giantswarm/issues/11260

Adding a new error matcher for validation issues.

### CODE REVIEW

File: service/controller/chart/resource/release/create.go [hunk 1]
```diff
@@ -103,13 +103,21 @@ func (r *Resource) ApplyCreateChange(ctx context.Context, obj, createChange inte
...
+		if helmclient.IsValidationFailedError(err) {
 			reason := err.Error()
 			reason = fmt.Sprintf("helm validation error: (%s)", reason)
 			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("helm release %#q failed, %s", releaseState.Name, reason))
 			addStatusToContext(cc, reason, validationFailedStatus)
 
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			resourcecanceledcontext.SetCanceled(ctx)
+			return nil
+		} else if helmclient.IsInvalidManifest(err) {
```
Comment (tomahawk28) [inline service/controller/chart/resource/release/create.go:115]: This is needed for some cases, for example when CRD is not there yet.

File: service/controller/chart/resource/release/update.go [hunk 2]
```diff
@@ -111,12 +111,21 @@ func (r *Resource) ApplyUpdateChange(ctx context.Context, obj, updateChange inte
 			return nil
 		}
 
-		if helmclient.IsInvalidManifest(err) {
+		if helmclient.IsValidationFailedError(err) {
```
Comment (tomahawk28) [inline service/controller/chart/resource/release/update.go:114]: Checking validation first, and later for invalid manifest.

**Comment by tomahawk28** [comment]
Seeing it failed to fill the right value for `status` resource
```
D 05/29 11:29:17 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/g8s-cert-manager-unique release.ApplyUpdateChange helm release `g8s-cert-manager-unique` failed, helm validation error: (error validating "": error validating data: ValidationError(RoleBinding.roleRef): unknown field "namespace" in io.k8s.api.rbac.v1beta1.RoleRef) | chart-operator/service/controller/chart/resource/release/update.go:117 | controller=chart-operator-chart | event=update | loop=48 | version=163056938
D 05/29 11:29:17 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/g8s-cert-manager-unique release.ApplyUpdateChange canceling resource | chart-operator/service/controller/chart/resource/release/update.go:120 | controller=chart-operator-chart | event=update | loop=48 | version=163056938
D 05/29 11:29:17 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/g8s-cert-manager-unique status getting status for release `g8s-cert-manager-unique` | chart-operator/service/controller/chart/resource/status/create.go:27 | controller=chart-operator-chart | event=update | loop=48 | version=163056938
D 05/29 11:29:18 /apis/application.giantswarm.io/v1alpha1/namespaces/giantswarm/charts/g8s-cert-manager-unique status setting status for release `g8s-cert-manager-unique` status to `` | chart-operator/service/controller/chart/resource/status/create.go:103 | controller=chart-operator-chart | event=update | loop=48 | version=163056938
```

**Comment by tomahawk28** [comment]
After the last commit, it works great. 
```
status:
  appVersion: 0.14.1
  reason: 'Helm reason: Upgrade complete Operator reason: helm validation error: (error
    validating "": error validating data: ValidationError(RoleBinding.roleRef): unknown
    field "namespace" in io.k8s.api.rbac.v1beta1.RoleRef)'
  release:
    lastDeployed: "2020-05-29T07:44:09Z"
    revision: 0
    status: validation-failed
  version: 1.1.1
```

File: service/controller/chart/resource/release/create.go (hunk 1 above)
Comment (rossf7) [reply service/controller/chart/resource/release/create.go:115]: Does the order matter or can we have this error first so they are ordered?

File: service/controller/chart/resource/release/update.go [hunk 3]
```diff
@@ -111,12 +111,21 @@ func (r *Resource) ApplyUpdateChange(ctx context.Context, obj, updateChange inte
...
 			reason = fmt.Sprintf("helm validation error: (%s)", reason)
 			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("helm release %#q failed, %s", releaseState.Name, reason))
 			addStatusToContext(cc, reason, validationFailedStatus)
 
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			resourcecanceledcontext.SetCanceled(ctx)
+			return nil
+		} else if helmclient.IsInvalidManifest(err) {
+			reason := err.Error()
+			reason = fmt.Sprintf("Invalid manifest error: (%s)", reason)
```
Comment (rossf7) [inline service/controller/chart/resource/release/update.go:125]:
```suggestion
			reason = fmt.Sprintf("invalid manifest error: (%s)", reason)
```

Can we make this lower case for consistency with the other messages?

File: service/controller/chart/resource/release/create.go (hunk 1 above)
Comment (tomahawk28) [reply service/controller/chart/resource/release/create.go:115]: Validation error during install starts with `unable to build kubernetes objects from release manifest: error validating data... ` but the same error during upgrade start as `error validating data...`. So I think current ordering is correct.

File: service/controller/chart/resource/release/update.go (hunk 3 above)
Comment (tomahawk28) [reply service/controller/chart/resource/release/update.go]: Thx, I fixed it for both create/update.go

File: service/controller/chart/resource/release/create.go (hunk 1 above)
Comment (rossf7) [reply service/controller/chart/resource/release/create.go:115]: OK makes sense 👍

**Pull request merged**